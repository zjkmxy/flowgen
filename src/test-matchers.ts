import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
