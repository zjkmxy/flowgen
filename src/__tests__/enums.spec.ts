import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle empty enums", async () => {
  const ts = `enum Empty { }`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle basic enums", async () => {
  const ts = `enum Label {
    LABEL_OPTIONAL,
    LABEL_REQUIRED,
    LABEL_REPEATED,
  }
type A = Label
type B = Label.LABEL_OPTIONAL
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle number enums", async () => {
  const ts = `enum Label {
    ONE = 1,
    TWO = 2,
    THREE = 3,
    NEGATIVE = -123,
    DECIMAL = 3.14,
  }
type A = Label
type B = Label.TWO
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle string enums", async () => {
  const ts = `enum Label {
    LABEL_OPTIONAL = 'LABEL_OPTIONAL',
    LABEL_REQUIRED = 'LABEL_REQUIRED',
    LABEL_REPEATED = 'LABEL_REPEATED',
  }
type A = Label
type B = Label.LABEL_REQUIRED
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle importing enum types", async () => {
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/export-enum-file.ts",
      "src/__tests__/snippet/import-enum-type-file.ts",
    ],
    {
      quiet: false,
    },
  );
  for (const result of results) {
    expect(await beautify(result[1])).toMatchSnapshot("class");
    // TODO: this function only runs flow on one file at a time, so it errors when trying to import
    // expect(result[1]).toBeValidFlowTypeDeclarations();
  }
});

it("should handle importing enums", async () => {
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/export-enum-file.ts",
      "src/__tests__/snippet/import-enum-file.ts",
    ],
    {
      quiet: false,
    },
  );
  for (const result of results) {
    expect(await beautify(result[1])).toMatchSnapshot("class");
    // TODO: this function only runs flow on one file at a time, so it errors when trying to import
    // expect(result[1]).toBeValidFlowTypeDeclarations();
  }
});

it("should handle import enum values", async () => {
  const ts = `enum Label {
    LABEL_OPTIONAL = 'LABEL_OPTIONAL',
    LABEL_REQUIRED = 'LABEL_REQUIRED',
    LABEL_REPEATED = 'LABEL_REPEATED',
  }
import B = Label.LABEL_REQUIRED
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});
