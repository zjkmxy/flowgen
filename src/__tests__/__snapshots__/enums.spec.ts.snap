// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle basic enums: class 1`] = `
"declare enum Label {
  LABEL_OPTIONAL = 0, // 0
  LABEL_REQUIRED = 1, // 1
  LABEL_REPEATED = 2, // 2
}
declare type A = Label;
declare type B = typeof Label.LABEL_OPTIONAL;
"
`;

exports[`should handle empty enums: class 1`] = `
"declare enum Empty {}
"
`;

exports[`should handle import enum values: class 1`] = `
"declare enum Label {
  LABEL_OPTIONAL = "LABEL_OPTIONAL", // "LABEL_OPTIONAL"
  LABEL_REQUIRED = "LABEL_REQUIRED", // "LABEL_REQUIRED"
  LABEL_REPEATED = "LABEL_REPEATED", // "LABEL_REPEATED"
}
var B = Label.LABEL_REQUIRED;
"
`;

exports[`should handle importing enum types: class 1`] = `
"declare export enum Label {
  A = "A", // "A"
  B = "B", // "B"
}
"
`;

exports[`should handle importing enum types: class 2`] = `
"import typeof { Label } from "./export-enum-file";
declare export function foo(label: Label): void;
"
`;

exports[`should handle importing enums: class 1`] = `
"declare export enum Label {
  A = "A", // "A"
  B = "B", // "B"
}
"
`;

exports[`should handle importing enums: class 2`] = `
"import { Label } from "./export-enum-file";
declare export function foo(label: Label): void;
"
`;

exports[`should handle number enums: class 1`] = `
"declare enum Label {
  ONE = 1, // 1
  TWO = 2, // 2
  THREE = 3, // 3
  NEGATIVE = -123, // -123
  DECIMAL = 3.14, // 3.14
}
declare type A = Label;
declare type B = typeof Label.TWO;
"
`;

exports[`should handle string enums: class 1`] = `
"declare enum Label {
  LABEL_OPTIONAL = "LABEL_OPTIONAL", // "LABEL_OPTIONAL"
  LABEL_REQUIRED = "LABEL_REQUIRED", // "LABEL_REQUIRED"
  LABEL_REPEATED = "LABEL_REPEATED", // "LABEL_REPEATED"
}
declare type A = Label;
declare type B = typeof Label.LABEL_REQUIRED;
"
`;
